stratum/lib/p4runtime/p4runtime_session.h

```diff
@@ -0,0 +1,152 @@
+
+  ~P4RuntimeSession();
+
+  // Disable copy semantics.
+  P4RuntimeSession(const P4RuntimeSession&) = delete;
+  P4RuntimeSession& operator=(const P4RuntimeSession&) = delete;
+
+  // Allow move semantics.
+  P4RuntimeSession(P4RuntimeSession&&) = default;
+  P4RuntimeSession& operator=(P4RuntimeSession&&) = default;
+
+  // Return the id of the node that this session belongs to.
+  uint32_t DeviceId() const { return device_id_; }
+  // Return the election id that has been used to perform master arbitration.
+  p4::v1::Uint128 ElectionId() const { return election_id_; }
+  // Return the P4Runtime stub.
+  p4::v1::P4Runtime::Stub& Stub() { return *stub_; }
+  // Reads back stream message response.
+  ABSL_MUST_USE_RESULT bool StreamChannelRead(
+      p4::v1::StreamMessageResponse* response) {
+    return stream_channel_->Read(response);
+  }
+  // Writes stream message request.
+  ABSL_MUST_USE_RESULT bool StreamChannelWrite(
+      const p4::v1::StreamMessageRequest& request) {
+    return stream_channel_->Write(request);
+  }
+  // Sets the forwarding pipeline from the given p4 info and device config.
+  ::util::Status SetForwardingPipelineConfig(
+      const p4::config::v1::P4Info& p4info,
+      const std::string& p4_device_config);
+  // Gets the current forwarding pipeline from the switch.
+  ::util::Status GetForwardingPipelineConfig(p4::config::v1::P4Info* p4info,
+                                             std::string* p4_device_config);
+  // Cancels the StreamChannel RPC. It is done in a best-effort fashion.
+  void TryCancel() { stream_channel_context_->TryCancel(); }
+
+ private:
+  P4RuntimeSession(uint32_t device_id,
+                   std::unique_ptr<p4::v1::P4Runtime::Stub> stub,
+                   absl::uint128 election_id)
+      : device_id_(device_id),
+        stub_(std::move(stub)),
+        stream_channel_context_(absl::make_unique<grpc::ClientContext>()),
+        stream_channel_(stub_->StreamChannel(stream_channel_context_.get())) {
+    election_id_.set_high(absl::Uint128High64(election_id));
+    election_id_.set_low(absl::Uint128Low64(election_id));
+  }
+
+  // The id of the node that this session belongs to.
+  uint32_t device_id_;
+  // The election id that has been used to perform master arbitration.
+  p4::v1::Uint128 election_id_;
+  // The P4Runtime stub of the switch that this session belongs to.
+  std::unique_ptr<p4::v1::P4Runtime::Stub> stub_;
+  // This stream channel and context are used to perform master arbitration,
+  // but can afterwards also be used for packet IO.
+  std::unique_ptr<grpc::ClientContext> stream_channel_context_;
+  std::unique_ptr<grpc::ClientReaderWriter<p4::v1::StreamMessageRequest,
+                                           p4::v1::StreamMessageResponse>>
+      stream_channel_;
+};
+
+// Creates P4Runtime stub with appropriate channel configuration.
+std::unique_ptr<p4::v1::P4Runtime::Stub> CreateP4RuntimeStub(
```

Comment:
We use this function only in one place, can move codes from this function to the `Create` function to avoid creating a shallow module

Reply:
You're right, as far as I can tell this code is structured with unit tests in mind. PINS has `p4_runtime_session_test.cc`. I'd include that here too, but it requires generating gRPC mocks in the p4runtime repo.

Reply:
`P4RuntimeSession::Create` can take a `::p4::v1::P4Runtime::MockStub` as input in unit tests.

`CreateP4RuntimeStub` is the public "production" interface with full setup.

Reply:
After some digging, I this is how it works:

`CreateP4RuntimeStub` is used on other tests that interact with P4runtime, but don't want an actual `P4RuntimeSession` with networking and so on.

`P4RuntimeSession::Create` is the interface production apps use. One-shot, batteries included.

Reply:
One more point: if we keep the code similar to the upstream PINS sources, we'll have an easier time integrating future updates.

Reply:
sounds good to me